### PR TITLE
Added an input flag to switch off multi-body hydrodynamic interactions

### PR DESCRIPTION
**Pull request by torrewk**

### CODE REVIEW

File: JFSD.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
 import numpy as np
 
-from jfsd import main, utils
+from jfsd import main, mainBD, utils
```
Comment (qubixes) [inline JFSD.py:3]: Module names should not have capital letters. I haven't changed anything for the current ones, so that I wouldn't break anything, but for new models we should not break the "rules".

File: jfsd/appliedForces.py [hunk 2]
```diff
@@ -206,7 +209,8 @@ def compute_hs_forces(
 
         #compute forces for each pair
         # relax constant (If lubrication is not on, k should to be ~ o(1) )
-        k = 1000.
+        k = jnp.where(HIs_flag > 0, 1000. , 100)
```
Comment (qubixes) [inline jfsd/appliedForces.py:212]: You might want to explain this / write a comment here why this is the case.

File: jfsd/main.py [hunk 3]
```diff
@@ -106,6 +110,8 @@ def main(
         Array of external forces (N,3)
     constant_applied_torques:
         Array of external torques (N,3)
+    HIs_flag:
```
Comment (qubixes) [inline jfsd/main.py:113]: Ideally, this flag would be a boolean. Is there a reason to use integers?

File: jfsd/mainBD.py
Comment (qubixes) [inline jfsd/mainBD.py:1]: I feel like there might be quite a bit of overlap between this and the normal main.py. Do you think parts could be reused instead of copied?

File: jfsd/utils.py [hunk 4]
```diff
@@ -717,3 +758,53 @@ def precompute(
     return ((all_indices_x), (all_indices_y), (all_indices_z), gaussian_grid_spacing1, gaussian_grid_spacing2,
             r, indices_i, indices_j, f1, f2, g1, g2, h1, h2, h3,
             r_lub, indices_i_lub, indices_j_lub,ResFunc)
+                                    
+@partial(jit, static_argnums=[3])
+def precomputeBD(
+        positions: float,
```
Comment (qubixes) [inline jfsd/utils.py:764]: If it is an array, then it should be an array of floats. If it were a list it would be `list[float]`.

**Review by qubixes — CHANGES_REQUESTED** [review]
I have looked through the parts that I can reasonably understand. The very technical parts are a little difficult to wrap my head around, especially the strides that seem to occur are pretty difficult. Hope my comments are helpful!

File: jfsd/main.py (hunk 3 above)
Comment (torrewk) [reply jfsd/main.py:113]: In future it will be possible to switch among 4 different hydrodynamic descriptions, thus the integer type.

Comment (qubixes) [reply jfsd/main.py:113]: Perhaps an Enum type would be more suitable then?

https://docs.python.org/3/library/enum.html

File: JFSD.py (hunk 1 above)
Comment (torrewk) [reply JFSD.py]: the module mainBD is not needed anymore [https://github.com/torrewk/Python-Jax-Fast-Stokesian-Dynamics/pull/13/commits/581bca0f2aa873674b5fa30f0f7d308d711c5fe1](url)

File: jfsd/appliedForces.py (hunk 2 above)
Comment (torrewk) [reply jfsd/appliedForces.py:213]: Added a more extensive comment [https://github.com/torrewk/Python-Jax-Fast-Stokesian-Dynamics/pull/13/commits/581bca0f2aa873674b5fa30f0f7d308d711c5fe1](url)

File: jfsd/mainBD.py
Comment (torrewk) [reply jfsd/mainBD.py:1]: Indeed, now there is only one main module

File: jfsd/utils.py (hunk 4 above)
Comment (torrewk) [reply jfsd/utils.py]: Type hints have been updated.

File: tests/test_class.py [hunk 5]
```diff
@@ -51,24 +89,46 @@ def theoretical_MSD_hashimoto(tempo, L):
...
+                2,0,0,0)
             msd.compute(positions=(traj))
             MSDs.append(msd.msd)
         MSDs_average = np.mean(MSDs, axis=0)
         error = np.abs((theoretical_MSD_hashimoto(
             np.arange(0, 100) * 0.1, 50))[:10] - MSDs_average[:10])
         assert (np.max(error) < 0.2)  # reference tolerance
+        
+    def test_thermal_realspace(self):
+        """Physical unit test for non-deterministic part of hydrodynamic calculations. Compare square root of resistance/mobility obtained
```
Comment (qubixes) [inline tests/test_class.py:116]: Doc strings should have a single line description, then one empty line, and then you can optionally add a more detailed description.

File: tests/test_class.py [hunk 6]
```diff
@@ -2,47 +2,85 @@
...
+            0., 1, 0.5, 0.001, 0., 0, 1.,
+            jnp.array([[0., 1.+dr, 0.], [0., -1.-dr, 0.]]),
+            0, 0, 0, 0, 0.1, 0.,
+            'None', 0, 0, 0,np.array([0]), np.array([0]),
+            1,0,0,0)
+        error = np.linalg.norm(reference_traj-traj)/np.linalg.norm(reference_traj)
+        assert (error < 0.17)
 
     def test_thermal_1body(self):
+        """Physical unit test for non-deterministic part of hydrodynamic calculations. Test Brownian motion of a single particle, in periodic boundary conditions.
```
Comment (qubixes) [inline tests/test_class.py:78]: Same as comment below, split the docstring.

**Review by qubixes — APPROVED** [review]
I have looked through the code for as much as I could, but honestly, I won't be able to understand its correctness by simply looking at it. You have checked the outcomes of the tests I presume?

File: tests/test_class.py (hunk 6 above)
Comment (torrewk) [reply tests/test_class.py]: Fixed

Comment (torrewk) [reply tests/test_class.py]: Added type and array dimension in description of each function
